virtual/activations.go

```diff
@@ -382,6 +383,46 @@ func (a *activations) getServerState() (
 	return a.serverState.serverID, a.serverState.serverVersion
 }
 
+func (a *activations) close(ctx context.Context, numWorkers int) error {
+	a.Lock()
+	defer a.Unlock()
+
+	// create a channel to receive keys
+	keys := make(chan types.NamespacedActorID, len(a._actors))
+	for k := range a._actors {
+		keys <- k
+	}
+	close(keys)
+
+	// create a wait group to wait for all goroutines to finish
+	g, ctx := errgroup.WithContext(ctx)
+
+	// spawn multiple goroutines to process actors shutdown
+	for i := 0; i < numWorkers; i++ {
+		g.Go(func() error {
+			for actorId := range keys {
+				// Process data[k]
+				future := a._actors[actorId]
+				// do something with d.value
+				actor, err := future.Wait()
+				if err != nil {
+					return fmt.Errorf("failed to wait for an actor: %w", err)
+				}
+
+				if err := actor.close(ctx); err != nil {
+					return fmt.Errorf("failed to close an actor: %w", err)
+				}
+
+				delete(a._actors, actorId)
+			}
+			return nil
+		})
+	}
+
+	// wait for all goroutines to finish
+	return g.Wait()
```

Comment:
I plan on returning an error in case we do not close all actors. And if so, I log the expected vs. closed actors number in error. And above, I log whenever an actor fails to be closed.